doc/nimgrep_cmdline.txt

```diff
@@ -0,0 +1,114 @@
+
+Usage:
+
+* To search::
```

Comment:
I understand you use `::` to make rst2html happy, but this now makes `nimgrep -h` show `::`

I completely agree with de-duplicating the help message, but we should post-process this file so it renders ok, eg:

```nim
proc postProcessRst(a: string): string =
  # TODO: handle `::`, etc
proc genHelp(): string =
  result = Version & """
  (c) 2012-2020 Andreas Rumpf
""" & slurp "../doc/nimgrep_cmdline.txt" 
  result = result.postProcessRst

const Usage = genHelp()
```

Reply:
I have a pr I've been working on for nimgrep. I deleted the old rst file and wrote this to keep it dry::
```nim

proc staticParseVerAndUsageFromDoc: (string, string) {.compileTime.} =
  ## Reads in this source file & parses out the title, version, author,
  ## copyright and cli help info from module's doc.
  ## There are many options to `nimgrep` and the old docs diverged over
  ## the years.
  template slice(offset: int, mark: string, endMark = "\n"): untyped =
    let off0 = source.find(mark, offset) + mark.len
    doAssert off0 > -1 + mark.len, "(" & mark.escape & " .. " & endMark.escape & ")"
    let off1 = source.find(endMark, off0) - 1
    doAssert off1 > -1, "(" & mark.escape & " .. " & endMark.escape & ")"
    off0..off1
  let source = staticRead currentSourcePath
  const doc0Mark = "\n\n## ="
  let doc0 = source.find(doc0Mark)
  doAssert doc0 > -1
  let title = slice(doc0 + doc0Mark.len, "=\n## ")
  let author = slice(title.b, "## :Author: ")
  let copyright = slice(author.b, "## :Copyright: ")
  let version = slice(copyright.b, "## :Version: ")
  var usage = slice(version.b, "\n## Usage\n## =", "\n## Design Notes\n## =")
  usage.a -= "\n## Usage\n## =".len
  result[0] = source[version]
  result[1] = "\n" & source[title] &
              "\n\n  Version " & source[version] &
              "\n  (c) " & source[copyright] & " " & source[author] &
              "\n" & source[usage]
  result[1] = result[1].multiReplace([("\n## ", "\n  "), ("\n##", "\n"),
                                      ("::\n", ":\n")])

const (Version, Usage) = staticParseVerAndUsageFromDoc()

```

Reply:
That of course was after I reformatted and added more docs such as a "Design Notes" section that was used the boundry of the cli usage sections.

Reply:
@quantimnot do you mean you put the whole `nimgrep.rst` inside `nimgrep.nim` as a doc comment?

Reply:
Yes. It was really outdated compared to the source. So I deleted it and added lots of docs to the source.

I was exploring other changes to nimgrep. [my nimgrep wip branch](https://github.com/quantimnot/Nim/blob/nimgrep/tools/nimgrep.nim)

Reply:
This approach makes sense to avoid documentation getting out of sync but I'm not sure that doc comments are appropriate for writing User's  Manual.

The 2 problems with it:
1. `*.nim` files get larger, though we should probably move some code to separate file(s) anyway
2. we may want to have internal (developer) documentation that can be generated by `nim doc --docInternal`. And this approach mixes Manual and Developer docs together.

May be it makes sense to introduce a next level of doc comments `###` specifically for writing User's Manuals, Guides, etc.

@timotheecour wdyt?

Reply:
There is nothing wrong with having nimgrep's documnetation in the `nimgrep.rst` file. Having this in the source code instead is not acceptable. Yes, yes, maybe it makes it slowly harder to change nimgrep's command line interface as the documentation then also needs to be updated. But that's a good thing -- some users trust nimgrep's command line **interface** not to change all the time...

Reply:
@Araq 

So what do you think about **this** PR? Say clearly please:

- yes, it's good to put cmdline options into `nimgrep_cmdline.txt` for both `nimgrep.nim` and `nimgrep.rst`
- no, it's better to have the options described separately even at the cost of duplication

Reply:
I can't speak for araq but as far as I'm concerned, the way you did it by introducing `nimgrep_cmdline.txt` (included by both nimgrep.nim and nimgrep.rst) is indeeed the way to go; there's no need for duplication here. And yes, it also allows having the documentation in a separate nimgrep.rst file.

The only thing is it needs a (hopefully simple) post-processing so that it renders well in both cmdline output and rst docs, but that can be done in a followup PR if needed. The post-processing can be reused in all similar instances and doesn't have to be specific to nimgrep.

Reply:
I agree with @timotheecour